File: cli/backend/book-generator.js
const { rename } = require("fs");
const { resolve: pathResolve } = require("path");
const bookEpubGenerator = require("./epub");
const { getBookIndex } = require("./helpers");
const BottomInfo = require("./bottom-info");
const bookAudioGenerator = require("./audio");
const { getAppDataPath } = require("./cross-platform");

const bookGenerator = (book) => {
    return new Promise((resolve, reject) => {
        const bInfo = new BottomInfo("Generating", "", ["📁", "⚡", "🧲", "📦️"]);
        bInfo.start();

        const info = getBookIndex(book, "info");
        if (info === false) {
            reject(new Error("Can not parse info json file!"));
            bInfo.end();
            return;
        }

        if (["mp3"].includes(info.type)) {
            bookAudioGenerator(book, info)
                .then((res) => {
                    bInfo.end();
                    resolve(res);
                })
                .catch((error) => {
                    bInfo.end();
                    reject(error);
                });
        } else if (info.type === "pdf") {
            const newPath = pathResolve(getAppDataPath("jarir-cli"), "books", book.title + "." + info.type);

            try {
                renameSync(pathResolve(getAppDataPath("jarir-cli"), "books", "" + book.id, "Text", "DATA.DATAx"), newPath);
            } catch (e) {
                bInfo.end();
                if (err) {
                    reject(err);
                }
            }
            resolve(newPath);
        } else if (info.type === "epub") {
            bookEpubGenerator(book, info)
                .then((res) => {
                    bInfo.end();
                    resolve(res);
                })
                .catch((err) => {
                    bInfo.end();
                    reject(err);
                });
        } else {
            bInfo.end();
            reject(new Error("file type: " + info.type + " is not supported yet!"));
        }
    });
};

module.exports = bookGenerator;

File: src-tauri/src/backend/book_generator.rs
use crate::backend::audio::book_audio_generator;
use crate::backend::book::Book;
use crate::backend::cross_platform::get_app_data_path;
use crate::backend::epub::book_epub_generator;
use crate::backend::helpers::get_book_index;
use std::path::PathBuf;
use thiserror::Error;
use tokio::fs::rename;

#[derive(Debug, Error)]
pub enum BookGeneratorError {
    #[error("Failed to parse info JSON file")]
    ParseError,
    #[error("File type {0} is not supported yet")]
    UnsupportedFileType(String),
    #[error("IO Error: {0}")]
    IoError(#[from] std::io::Error),
    #[error("Tokio Join Error: {0}")]
    JoinError(#[from] tokio::task::JoinError),
}

pub async fn book_generator(book: Book) -> Result<PathBuf, BookGeneratorError> {
    let info: serde_json::Value =
        get_book_index(&book.id, "info").ok_or(BookGeneratorError::ParseError)?;

    let book_type = info["type"].as_str().unwrap_or_default();

    match book_type {
        "mp3" => {
            let res = book_audio_generator(book, Some(info)).await;
            Ok(res.unwrap())
        }
        "pdf" => {
            let new_path = get_app_data_path(Some("books"))
                .join(format!(
                    "{}.{}",
                    book.title,
                    info["r#type"].as_str().unwrap_or_default()
                ));
            let old_path = get_app_data_path(Some("books"))
                .join(&book.id)
                .join("Text")
                .join("DATA.DATA");

            rename(old_path, &new_path).await?;
            Ok(new_path)
        }
        "epub" => {
            let res = book_epub_generator(book, Some(info)).await;
            Ok(res.unwrap())
        }
        _ => Err(BookGeneratorError::UnsupportedFileType(
            book_type.to_string(),
        )),
    }
}
